### PR TITLE
add hasWarnings handler

### PR DESCRIPTION
**Pull request by lewis150193**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/RWA-239

### Change description ###
Added warn handler

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ x ] No
```

### CODE REVIEW

File: src/test/java/uk/gov/hmcts/reform/wacaseeventhandler/services/WarningTaskHandlerTest.java [hunk 1]
```diff
@@ -1,30 +1,127 @@
...
+                .fromState(new DmnStringValue("some previous state"))
+                .build();
+
+        EvaluateDmnRequest<CancellationEvaluateRequest> requestParameters =
+            new EvaluateDmnRequest<>(cancellationTaskEvaluateDmnRequestVariables);
+
+        when(workflowApiClientToWarnTask.evaluateDmn(
+            DMN_NAME,
+            requestParameters
+        )).thenReturn(new EvaluateDmnResponse<>(Collections.emptyList()));
```
Comment (raghera) [inline src/test/java/uk/gov/hmcts/reform/wacaseeventhandler/services/WarningTaskHandlerTest.java:73]: Probably worth having a test where the response is returning values also and then can assert that the same ones are returned.

File: src/test/java/uk/gov/hmcts/reform/wacaseeventhandler/services/WarningTaskHandlerTest.java [hunk 2]
```diff
@@ -1,30 +1,127 @@
...
+
+        handlerService.evaluateDmn(eventInformation);
```
Comment (raghera) [inline src/test/java/uk/gov/hmcts/reform/wacaseeventhandler/services/WarningTaskHandlerTest.java:75]: should check this is an empty list as expected.

File: src/test/java/uk/gov/hmcts/reform/wacaseeventhandler/services/WarningTaskHandlerTest.java [hunk 3]
```diff
@@ -1,30 +1,127 @@
...
+        .newStateId("some post state")
+        .previousStateId("some previous state")
+        .jurisdictionId("ia")
+        .caseTypeId("asylum")
+        .caseReference("some case reference")
+        .dateTime(LocalDateTime.now())
+        .build();
 
     @Test
     void evaluateDmn() {
```
Comment (raghera) [inline src/test/java/uk/gov/hmcts/reform/wacaseeventhandler/services/WarningTaskHandlerTest.java:59]: I'd recommend having more descriptive names for tests... e.g. `should_call_evaluateDmn_and_return_empty_list_response`

The idea being that someone can just read the test title to know what's going on in it.

File: src/test/java/uk/gov/hmcts/reform/wacaseeventhandler/services/WarningTaskHandlerTest.java [hunk 4]
```diff
@@ -1,30 +1,127 @@
...
     }
 
+
     @Test
-    void handle() throws NoSuchMethodException {
-        // change for something meaningful once Warning handler is implemented
-        assertThat(handlerService.getClass().getMethod("evaluateDmn", EventInformation.class).getName())
-            .isEqualTo("evaluateDmn");
+    void handle() {
+        WarningEvaluateResponse result1 = WarningEvaluateResponse.builder()
```
Comment (raghera) [inline src/test/java/uk/gov/hmcts/reform/wacaseeventhandler/services/WarningTaskHandlerTest.java:87]: Probably a good idea to add more non-warning EvaluateResponses e.g. Cancellation ones.  Since the SUT is filtering them out.  Then this test can check this is working correctly.

File: src/test/java/uk/gov/hmcts/reform/wacaseeventhandler/services/WarningTaskHandlerTest.java [hunk 5]
```diff
@@ -1,30 +1,127 @@
...
+            .action(new DmnStringValue("Warn"))
+            .build();
+
+        WarningEvaluateResponse result2 = WarningEvaluateResponse.builder()
+            .action(new DmnStringValue("Warn"))
+            .build();
+
+        List<WarningEvaluateResponse> results = List.of(result1, result2);
```
Comment (raghera) [inline src/test/java/uk/gov/hmcts/reform/wacaseeventhandler/services/WarningTaskHandlerTest.java:95]: You can also have a test case where there is an empty list of results being sent.  It can check that then there is no interaction with the sendMessage endpoint.

File: src/main/java/uk/gov/hmcts/reform/wacaseeventhandler/clients/WorkflowApiClientToWarnTask.java [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+            String.format("%s/workflow/decision-definition/key/%s/tenant-id/%s/evaluate", workflowApiUrl, key,tenantId),
+            HttpMethod.POST,
+            new HttpEntity<>(requestParameters, buildHttpHeader()),
+            new ParameterizedTypeReference<>() {
+            }
+        ).getBody();
+    }
+
+    @Override
+    public ResponseEntity<Void> sendMessage(
```
Comment (BenNzewi) [inline src/main/java/uk/gov/hmcts/reform/wacaseeventhandler/clients/WorkflowApiClientToWarnTask.java:61]: test to cover this method?

File: src/integrationTest/java/uk/gov/hmcts/reform/wacaseeventhandler/controllers/CaseEventHandlerControllerTest.java [hunk 7]
```diff
@@ -9,7 +9,7 @@
 import uk.gov.hmcts.reform.wacaseeventhandler.domain.handlers.common.DmnStringValue;
 import uk.gov.hmcts.reform.wacaseeventhandler.domain.handlers.common.EventInformation;
 import uk.gov.hmcts.reform.wacaseeventhandler.domain.handlers.initiatetask.InitiateEvaluateResponse;
-import uk.gov.hmcts.reform.wacaseeventhandler.domain.handlers.warningtask.WarningEvaluateResponse;
+import uk.gov.hmcts.reform.wacaseeventhandler.domain.handlers.warningtask.WarningResponse;
```
Comment (BenNzewi) [inline src/integrationTest/java/uk/gov/hmcts/reform/wacaseeventhandler/controllers/CaseEventHandlerControllerTest.java:12]: Can we add Functional tests for the AC's in the ticket?
1. Given the makeAnApplication event has occurred
And the Cancellation DMN rules have run

2.Given the Warning Sub-process has been triggered 
And the Handle Warning service task has started
When the task in the parent process has attribute "hasWarnings" = No

3.Given the Warning Sub-process has been triggered 
When the task in the parent process has attribute "hasWarnings" = Yes
Then the Handle warning service task will make no change

File: src/functionalTest/java/uk/gov/hmcts/reform/wacaseeventhandler/controllers/CaseEventHandlerControllerTest.java [hunk 8]
```diff
@@ -139,6 +139,44 @@ public void given_initiated_tasks_with_delayTimer_toFuture_with_followup_overdue
         assertTaskDeleteReason(task1Id, "deleted");
     }
```
Comment (BenNzewi) [inline src/functionalTest/java/uk/gov/hmcts/reform/wacaseeventhandler/controllers/CaseEventHandlerControllerTest.java:141]: Can we add the test to cover AC for **makeAnApplication** event occurring and then warning message being received is not covered

File: src/main/java/uk/gov/hmcts/reform/wacaseeventhandler/handlers/WarningTaskHandler.java [hunk 9]
```diff
@@ -2,24 +2,86 @@
...
+        return new EvaluateDmnRequest<>(variables);
     }
 
     @Override
     public void handle(List<? extends EvaluateResponse> results, EventInformation eventInformation) {
-        // empty for now
+        results.stream()
+            .filter(result -> result instanceof WarningEvaluateResponse)
+            .map(result -> (WarningEvaluateResponse) result)
+            .forEach(cancellationEvaluateResponse -> sendMessageToWarnTasksForGivenCorrelations(
```
[the remark this answers is not in the document]
Comment (iguannago) [reply src/main/java/uk/gov/hmcts/reform/wacaseeventhandler/handlers/WarningTaskHandler.java]: Yes, I understand what you say, @raghera . However, the forEach iterates over the rows resulting from evaluating the Cancellation DMN (the warning rule is part of this table as it is the cancellation rules, yes). Therefore, yes, we can create a new object and name it as you say, but it's going to look exactly the same as the CancellationEvaluateResponse . Just wanted to clarify it. Having said, I agree with Ravi's change since it makes it clearer.

File: src/main/java/uk/gov/hmcts/reform/wacaseeventhandler/handlers/WarningTaskHandler.java [hunk 10]
```diff
@@ -2,24 +2,86 @@
...
+                eventInformation.getCaseReference(),
+                cancellationEvaluateResponse.getAction().getValue()
+            ));
+    }
+
+    private void sendMessageToWarnTasksForGivenCorrelations(String caseReference, String category) {
+        workflowApiClientToWarnTask.sendMessage(buildSendMessageRequest(category, caseReference));
```
[the remark this answers is not in the document]
Comment (iguannago) [reply src/main/java/uk/gov/hmcts/reform/wacaseeventhandler/handlers/WarningTaskHandler.java]: Yes, I think the initiate takes that into account. So we have an example to follow.

File: src/main/java/uk/gov/hmcts/reform/wacaseeventhandler/handlers/WarningTaskHandler.java [hunk 11]
```diff
@@ -2,24 +2,86 @@
...
+        results.stream()
+            .filter(result -> result instanceof WarningEvaluateResponse)
+            .map(result -> (WarningEvaluateResponse) result)
+            .forEach(cancellationEvaluateResponse -> sendMessageToWarnTasksForGivenCorrelations(
+                eventInformation.getCaseReference(),
+                cancellationEvaluateResponse.getAction().getValue()
+            ));
+    }
+
+    private void sendMessageToWarnTasksForGivenCorrelations(String caseReference, String category) {
```
[the remark this answers is not in the document]
Comment (iguannago) [reply src/main/java/uk/gov/hmcts/reform/wacaseeventhandler/handlers/WarningTaskHandler.java]: yes, I think it was done initially to have a name and make it clearer, but yeah, we can inline it.

**Review by iguannago — APPROVED** [review]
